piet-cairo/src/text.rs

```diff
@@ -134,37 +253,104 @@ impl TextLayoutBuilder for CairoTextLayoutBuilder {
+
+        debug_assert!(
+            range.start >= self.last_range_start_pos,
+            "attributes must be added in non-decreasing start order"
+        );
+        self.last_range_start_pos = range.start;
+
+        self.attributes.push(AttributeWithRange {
+            attribute,
+            range: Some(range),
+        });
+
         self
     }
 
     fn build(self) -> Result<Self::Out, Error> {
-        // set our default font
-        let font = CairoFont::new(self.defaults.font.clone());
-        let size = self.defaults.font_size;
-        let weight = if self.defaults.weight.to_raw() <= piet::FontWeight::MEDIUM.to_raw() {
-            FontWeight::Normal
-        } else {
-            FontWeight::Bold
-        };
-        let slant = match self.defaults.style {
-            FontStyle::Italic => FontSlant::Italic,
-            FontStyle::Regular => FontSlant::Normal,
+        let pango_attributes = AttrList::new();
+        let add_attribute = |attribute| {
+            if let Some(attribute) = attribute {
+                pango_attributes.insert(attribute);
+            }
         };
 
-        let scaled_font = font.resolve(size, slant, weight);
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::FontFamily(self.defaults.font),
+                range: None,
+            }
+            .into_pango(),
+        );
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::FontSize(self.defaults.font_size),
+                range: None,
+            }
+            .into_pango(),
+        );
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::Weight(self.defaults.weight),
+                range: None,
+            }
+            .into_pango(),
+        );
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::TextColor(self.defaults.fg_color),
+                range: None,
+            }
+            .into_pango(),
+        );
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::Style(self.defaults.style),
+                range: None,
+            }
+            .into_pango(),
+        );
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::Underline(self.defaults.underline),
+                range: None,
+            }
+            .into_pango(),
+        );
+        add_attribute(
+            AttributeWithRange {
+                attribute: TextAttribute::Strikethrough(self.defaults.strikethrough),
+                range: None,
+            }
+            .into_pango(),
+        );
+
+        for attribute in self.attributes {
+            add_attribute(attribute.into_pango());
+        }
+
+        self.pango_layout.set_attributes(Some(&pango_attributes));
+
+        //NOTE: We give Pango a width of -1 for no wrapping
```

Comment:
this is documented? In this case I would change above in where we set our default `wrap_width` to `f64::INFINITY`.

Reply:
Pango likes using -1 for "no width/height", I'll make this comment slightly better
See: https://gtk-rs.org/docs/pango/struct.Layout.html#method.set_width